### DISCUSSION TITLE
Issues installing/runnning BrainVisa on silicon Macs

### DISCUSSION THREAD

**Discussion by bobssh** [brainvisa/brainvisa.github.io#187] [Q&A] [answered]
Attempting to install BrainVisa on an m3 MacBook Pro, Ive encountered the following issues:

Apptainer install:
Appatiner itself installs fine but when running the .sif file I encounter the following error message.
"FATAL:   While checking container encryption: could not open image /Users/NAME/brainvisa-5.0.0/brainvisa-5.0.0.sif: the image's architecture (amd64) could not run on the host's (arm64)"
This error occurs with both the user and developer versions. I also tried to create an arm64 compatible image by pulling one from docker while specifying the architecture type as arm64, this successfully created a new .sif file but produced the same error message when ran. 

VBOX install:
The developers beta version of VBOX does run on silicon Macs however when attempting to start the BrainVisa virtual machine, it eventually crashes (I can provide the crash report). I also ran through the Mac troubleshooting, disabling audio etc. but this had the same outcome. 

Docker desktop:
Attempting to run the docker image on docker desktop (with amd64 emulation enabled through Rosetta), similarly caused it to crash. 

Is there anything else I could try or should this be taken as the program is incompatible with silicon? If so are there any plans to bring BrainVisa to silicon?
Thank you!

**Comment by sapetnioc** [comment]
First, thank you for your efforts in making BrainVISA work on Mac. The main problem is that no one in the BrainVISA team use Mac. Therefore we cannot test and debug and lack knowledge of this platform. Without external help, BrainVISA will remain incompatible with Silicon. Crash problems could be identified and even resolved by bringing together a BrainVISA expert and a motivated Mac expert. It is difficult to give you clues on what to do without knowing the reasons of the crashes (there could be a different reason in each environment). However, for this kind of system related problem, I usually not start by launching `brainvisa` because it starts various components and it is sometime difficult to know the one that has a problem. I first try to assess if a non graphical compiled command is working:
```
AimsFileInfo -h
```
If yes, can it read an image:
```
AimsFileInfo -v -i {image file}
```
If it works, I switch to `anatomist` because it uses both Qt for the GUI and OpenGL for rendering which are the main sources of incompatibilities with the system. If main Anatomist window opens it mean that Qt is probably fine and if I can open a 3D window, it means that OpenGL can work. Reaching this step usually means that BrainVISA will work.

You can try that and tell us the result. If you agree the next step could be to work with at least one of us sharing your screen so that we can put our knowledges in common to identify the problem.

**Reply by bobssh** [reply]
Thanks for your response, I have tried the commands you suggested but unfortunately they aren't recognised because I haven't been able to run the .sif file through Apptainer, so the setup procedure hasn't ran either. 

The issue with virtual box may well be a problem with their own silicon beta and not the Brainvisa container, for this reason I'm more inclined to try and get the singularity working. I haven't given up trying to use apple's rosetta emulation but have not had any luck yet. The main issue I'm facing with the singularity route is being unable to open the container because of its differing architecture type. I don't know how difficult or time intensive this would be but I believe there are methods to build cross-compatible containers without access to an arm64 machine through emulation (https://sylabs.io/2022/07/cross-architecture-containers-with-singularityce-pro/).

If that step could be passed, then any further issues could hopefully be worked on by sharing screen etc. Unfortunately I'm neither a Mac expert nor a Brainvisa expert, but I'm happy to try and help!

**Reply by sapetnioc** [reply]
I did no understand fully the problems with  AppTainer and VirtualBox. I don't think I can be of much help to fix a virtual environment on Mac. Being able to use an emulation for the architecture would be great because it does not require to change anything on BrainVISA side. I did not know it was even possible.

There is another possibility but is would require much more work on our side and there is a good chance that it is very complicated to setup : compile a Mac/arm64 version. This is a hazardous path because things can go wrong in many ways when using a new architecture. But if there is a possibility to setup an Ubuntu (maybe Debian is sufficient) distro in a container, we could try to setup a development environment just in case it work out of the box (I do not think so).

If it is not possible with a container, the last possibility is even more hazardous. We have an experimental development environment fully based on Conda packages. This environment do not use Conda but an alternative called (Pixi)[https://pixi.sh] that exists on MacOS. If all the dependencies we use on Linux are available on Mac, it may work.

I would definitely prefer a solution compatible with our x86_64 distribution but if you are desperate enough, I can guide you in the compilation steps. But I do not guaranty the result and, if it fail, I bet we will have a lot of complex interaction before we give up.

**Reply by bobssh** [reply, marked as answer]
A member of the computing staff at my university has found a solution! Essentially using a software called UTM to run the VirtualBox Image, this runs through a CPU emulation however I have found the performance to be quite good. 

I have attached a pdf copy of the instructions. Steps 1 & 2 are to extract VirtualBox disk image (.vmdk file) from the .ova file that you provide on the BrainVisa webpage. If you were to provide the .vmdk file on the webpage, then steps 1 & 2 could be ignored.


[Silicon_Mac_BrainVisa_Instructions.pdf](https://github.com/user-attachments/files/17072984/Silicon_Mac_BrainVisa_Instructions.pdf)

**Comment by denisri** [comment]
Hi,
Thank you very much for the instructions. We will include them in the download/install instructions on BrainVisa website. Is there any acknowledgement that we can add to it ?

**Comment by denisri** [comment]
We have added the instructions in https://brainvisa.info/web/download.html
It is still possible, of course, to add acknowledgements.
Thanks again,
Denis
